src/oic/extension/provider.py

```diff
@@ -640,19 +640,9 @@ def client_credentials_grant_type(self, areq):
         return Response(atr.to_json(), content="application/json")
 
     def password_grant_type(self, areq):
-        _at = self.token_handler.get_access_token(areq['client_id'],
```

Comment:
I don't see any kind of password / username check in here.

As we don't have a real user database to ask, there should probably be a call to some authn code somewhere in here, but i don't know how it would look. Probably just some callback to override by users of the class when they attach some form of authentication. 

The existing authn code isn't entirely useful here, as we do not know which authn method to pick (e.g. it might be a local password database or an LDAP based one, or both in parallel under different ACRs.), and we do not know how to pass username, password as the `authenticated_as()` methods do not have a standard for that.

Reply:
The credentials check are done in the `token_endpoint` method. This just handles the specifics. So we already have a verified client and we are just creating an access token based on that.

The user authentication is separate and was done in previous step.

Reply:
And as for the password in the request, I am using a "trick" with session identifier. The establishing of username/password is out of scope of the specs, and it would need to be handled manually anyway, so I am using the existing sessionDB there. Not sure if any way for easy customization is really necessary.

Reply:
I don't really see where the user authn happens in this case. Client authn happens in the token_endpoint, yes. 

But for the token_endpoint() no authn based on the user happens anywhere, at least not in the example code that uses it (https://github.com/OpenIDC/pyoidc/blob/master/oauth_example/as/as.py) that goes straight from request to the provider.token_endpoint().

Your trick obviously works for a trivial demo. But how would you add ROCP to the example above, e.g. to log in with 'diana'?

Reply:
As is said in 4.3.1, the method of obtaining the username/password is out of scope. The user authentication would be done there and the client credentials issued there as well.
This would require the user to rewrite/augment AuthorizationEndpoint to fit this job.

I guess that this brings an interesting question - what should we do with the extension module? We can probably either drop it completely (we don't need the code for OIDC) or integrate it into Oauth module (we will try to be a complete OAuth 2.0 implementation).

Reply:
4.3.1 talks about the client. The flow in 4.3 (C) talks about the server, which is the provider, and 4.3.2 explicitly says "The authorization server MUST:  
- require client authentication for confidential clients or for any client that was issued client credentials (or with other authentication requirements),
- authenticate the client if client authentication is included, and
- validate the resource owner password credentials using its existing password validation algorithm.
"
We do the first two here, but skip the third. I would be ok if the code called a method like 'verify_user_and_password' that simply raises NotImplementedError() in the default. It should return the uid of the authenticated user (a bit like the authn plugins). A user would need to add that part.

There are some interesting things in the extension module, that would be nice to have (more features for client registration for example, or the software statement/code stuff) so integrating it would probably be the best longer term thing to do.

Reply:
Ah, I see. I think I misunderstood the specs. Will update accordingly.